### PR TITLE
Cli demo

### PR DESCRIPTION
**Pull request by alari**

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -1,4 +1,5 @@
 .idea
 target
 log
-storage/data
\ No newline at end of file
+storage/data
+application.conf
```
Comment (RicoGit) [inline .gitignore]: application.conf should not be in the root folder of project. what is this line for?

File: client/src/main/scala/fluence/client/ClientComposer.scala [hunk 2]
```diff
@@ -49,4 +62,49 @@ object ClientComposer {
...
+      client ← fluenceClient(contacts, signAlgo, cryptoHasher, config)
+    } yield client
+
+  def getKeyPair(config: Config, algo: SignAlgo): IO[KeyPair] =
+    for {
+      kpConf ← KeyPairConfig.read(config)
+      kp ← FileKeyStorage.getKeyPair[IO](kpConf.keyPath, algo)
+    } yield kp
+
+  private def fluenceClient(
```
Comment (RicoGit) [inline client/src/main/scala/fluence/client/ClientComposer.scala]: May be better **create**FluenceClient?

File: client/src/main/scala/fluence/client/FluenceClient.scala [hunk 3]
```diff
@@ -130,19 +123,19 @@ class FluenceClient(
    * if it is no contract, create new contract and dataset
    *
    */
-  private def restoreDataset(ac: AuthorizedClient): Task[ClientDatasetStorageApi[Task, String, String]] = {
+  private def restoreDataset(kp: KeyPair): Task[ClientDatasetStorageApi[Task, String, String]] = {
```
Comment (RicoGit) [inline client/src/main/scala/fluence/client/FluenceClient.scala]: Why was it necessary to change AuthorizedClien to KeyPair? Tomorrow you will need to transfer the keys for encryption here. It looks like an useless change

File: client/src/main/scala/fluence/client/FluenceClient.scala [hunk 4]
```diff
@@ -209,66 +202,23 @@ class FluenceClient(
...
     kademliaClient: Kademlia[Task, Contact],
     contracts: Contracts[Task, BasicContract, Contact],
     storageRpc: Contact ⇒ DatasetStorageRpc[Task],
-    signAlgo: SignAlgo = Ecdsa.signAlgo,
-    storageHasher: CryptoHasher[Array[Byte], Array[Byte]]): FluenceClient = {
+    signAlgo: SignAlgo,
+    storageHasher: CryptoHasher[Array[Byte], Array[Byte]]): FluenceClient =
     new FluenceClient(kademliaClient, contracts, signAlgo, storageRpc, storageHasher)
-  }
+
```
Comment (RicoGit) [inline client/src/main/scala/fluence/client/FluenceClient.scala]: Are you sure it should be committed?

File: client/src/main/scala/fluence/client/cli/Cli.scala [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+          false
+        }
+
+      case Some(CliOp.Put(k, v)) ⇒
+        val t = for {
+          ds ← fluenceClient.getOrCreateDataset(kp)
+          _ ← ds.put(k, v)
+          _ = logger.info("Success.")
+        } yield true
+        t.toIO
```
Comment (RicoGit) [inline client/src/main/scala/fluence/client/cli/Cli.scala]: What is 't' mean? Do you save bytes of source code?

File: client/src/main/scala/fluence/client/cli/Cli.scala [hunk 6]
```diff
@@ -0,0 +1,71 @@
...
+import cats.effect.IO
+import fluence.client.FluenceClient
+import fluence.crypto.keypair.KeyPair
+import monix.execution.Scheduler.Implicits.global
+import org.jline.reader.LineReaderBuilder
+import org.jline.terminal.TerminalBuilder
+
+import scala.language.higherKinds
+
+object Cli extends slogging.LazyLogging {
```
Comment (RicoGit) [inline client/src/main/scala/fluence/client/cli/Cli.scala]: Documentation for this class is not needed?

File: client/src/main/scala/fluence/client/cli/CliInterpreter.scala [hunk 7]
```diff
@@ -0,0 +1,47 @@
...
+ */
+
+package fluence.client.cli
+
+import cats.effect.IO
+import cats.~>
+import org.jline.reader.LineReaderBuilder
+import org.jline.terminal.TerminalBuilder
+
+class CliInterpreter() extends (CliOp ~> IO) {
```
Comment (RicoGit) [inline client/src/main/scala/fluence/client/cli/CliInterpreter.scala]: Documentation for this class is not needed?

File: client/src/main/scala/fluence/client/cli/CliInterpreter.scala [hunk 8]
```diff
@@ -0,0 +1,47 @@
...
+
+  //for terminal improvements: history, navigation
+  private val terminal = TerminalBuilder.terminal()
+  private val lineReader = LineReaderBuilder.builder().terminal(terminal).build()
+
+  import CliOp._
+
+  override def apply[A](fa: CliOp[A]): IO[A] = ???
```
Comment (RicoGit) [inline client/src/main/scala/fluence/client/cli/CliInterpreter.scala]: If this class isn't finish and isn't works, why it should be here?

File: client/src/main/scala/fluence/client/cli/CliOps.scala [hunk 9]
```diff
@@ -15,24 +15,25 @@
...
-/**
- * A class that is an authorized user who can use datasets
- * @param kp a pair of keys with a public key that will be used as an address for dataset ids and contracts
- */
-case class AuthorizedClient(kp: KeyPair)
+class CliOps[F[_]](implicit I: InjectK[CliOp, F]) {
+
+  def exit: Free[F, Unit] = inject[CliOp, F](CliOp.Exit)
+
+  def get(key: String): Free[F, Unit] = inject[CliOp, F](CliOp.Get(key))
```
Comment (RicoGit) [inline client/src/main/scala/fluence/client/cli/CliOps.scala]: Are you discussed the removal AuthorizedClient with Dima?

File: client/src/main/scala/fluence/client/cli/Cli.scala [hunk 10]
```diff
@@ -0,0 +1,71 @@
...
+        IO {
+          logger.info("Exiting from fluence network.")
+          false
+        }
+
+      case Some(CliOp.Put(k, v)) ⇒
+        val t = for {
+          ds ← fluenceClient.getOrCreateDataset(kp)
+          _ ← ds.put(k, v)
+          _ = logger.info("Success.")
```
Comment (RicoGit) [inline client/src/main/scala/fluence/client/cli/Cli.scala]: Maybe we should show previous value here?

File: .gitignore (hunk 1 above)
Comment (alari) [reply .gitignore]: It adds all the `application.conf` files to git ignore

File: client/src/main/scala/fluence/client/cli/CliOps.scala (hunk 9 above)
Comment (DieMyst) [reply client/src/main/scala/fluence/client/cli/CliOps.scala]: Yes, we discussed it. We decided to use private key as password to simplify flow of cli usage

File: client/src/main/scala/fluence/client/ClientReplicationWrapper.scala [hunk 11]
```diff
@@ -64,13 +64,23 @@ class ClientReplicationWrapper[K, V](
...
-        logger.info(s"$key and $value was written to $replicationFactor nodes")
-        seq.head // at least one server should be here, we can't allocate contract without participants
-      }
+    for {
+      //check that all datasets is available, for demo purpose only
+      getRes ← Task.sequence(
+        datasetReplicas
+          .map { case (store, _) ⇒ store.get(key) }
+      )
+      _ ← if (getRes.size < datasetReplicas.size)
```
Comment (RicoGit) [inline client/src/main/scala/fluence/client/ClientReplicationWrapper.scala]: This code is not required. Task.sequence should raise error, when at least one task raise error

Comment (DieMyst) [reply client/src/main/scala/fluence/client/ClientReplicationWrapper.scala]: I checked this and if one node is offline, then it just return one result

**Pull request merged**